### DISCUSSION TITLE
Discussion: lock and using expressions

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#451] [General]
```cs
var res = lock (obj) M();
var res = using (var x = e) x.M();
```
instead of
```cs
string res;
lock (obj) { res = M(); }
using (var x = e) { res = x.M(); }
```
This lets you use type inferred locals which can also be made readonly (https://github.com/dotnet/csharplang/issues/188).
```cs
let res = lock (obj) M();
let res = using (var x = e) x.M();
```

**Comment by aluanhaddad** [comment]
`using`👍 
`lock` 👎

**Comment by alrz** [comment]
@aluanhaddad Any reasons why?

**Comment by CyrusNajmabadi** [comment]
I would def find value in a lock-expression.

**Comment by jnm2** [comment]
I've thought about lock expressions a lot recently. Using expressions not so much.

**Comment by FunctionalFirst** [comment]
You could take [a page out of the F# playbook](https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core/prim-types.fs#L4727-L4731) and do something like this:

```c#
static T Lock<T>(object syncObj, Func<T> fn) {
    lock (syncObj) {
        return fn();
    }
}

var obj = new object();
var res = Lock(obj, M);
```

**Comment by jnm2** [comment]
As long as that is the only time in your career where you invoke a delegate inside a lock statement, I promise not to have a heart attack. :grin:

**Comment by FunctionalFirst** [comment]
@jnm2 How is that any different than ``lock(obj) M();``?

**Comment by jnm2** [comment]
@FunctionalFirst Your code sample is fine! I'm wary of delegates inside a lock in general. It can lead to reentry issues or deadlocks in corner cases depending who gave you the delegate. For example, the delegate might raise an event.

**Comment by aluanhaddad** [comment]
@alrz 
> @aluanhaddad Any reasons why?

Now that I think about it, I don't know, both seem reasonable. I just feel like locks should be very rare and don't need a sugar.

**Comment by frederik-h** [comment]
One place where I really miss lock expressions is in expression bodied properties. I use those a lot, but it is not possible whenever I need to use a lock. I would like how lock expressions would make the code style more consistent.

**Comment by GeirGrusom** [comment]
Inconsistency is a bad trait of any programming language. If it works for `using` it should work for `lock` as well simply because it would be odd if it didn't.

**Comment by jnm2** [comment]
@GeirGrusom I used to think that, but I'm not actually sure why that's really an important factor anymore.

**Comment by GeirGrusom** [comment]
I hate stumbling upon things that don't work simply because it doesn't. SQL dialects are filled to the brim with those, and encountering inconsistencies doesn't exactly brighten my day. I never say "Oh how wise of the language designers to make that inconsistent with the language syntax". More commonly it makes me talk trash about that language at the water cooler.

**Comment by jnm2** [comment]
@GeirGrusom `throw default` is disallowed for example. `unsafe` can go in member and type declarations, but `checked` cannot. It's a matter of perspective; I'd say there's such a thing as a good inconsistency when the inconsistency puts friction in the right place like `throw default`. I'd have the same opinion about an expression form of `lock`; I think there should be friction to to encourage you to avoid having a lot of simple members that lock.

**Comment by alrz** [comment]
> there should be friction to to encourage you to avoid having a lot of simple members that lock.

I don't think that's useful. convinience in doing so when you need it isn't harmful and the "friction" wouldn't help you to overcome the requirement.

consider this example:

```cs
lock(o)
  M(CriticalMethod(), M());
```
here you have held the lock for uncritical sections as well. either you need to manually capture the result in a local,
```cs
int result;
lock (o)
  result = CriticalMethod();
M(result, M());
```
or have the compiler generate it for you,
```cs
M(lock (o) CriticalMethod(), M());
```
so you release the lock as soon as you're done with it without adding any complexity to the flow of the code.

**Comment by Joe4evr** [comment]
> `unsafe` can go in member and type declarations, but `checked` cannot

We need to bring back support for #552.

**Comment by jnm2** [comment]
Hehe https://github.com/dotnet/csharplang/issues/552#issuecomment-299937832

**Comment by Joe4evr** [comment]
I know, that's why I said **we**. :laughing:

**Comment by HaloFour** [comment]
@alrz 

You're last example doesn't work unless `lock (x) expression` is also considered an expression, which is a bigger change to the language than just allowing the removal of the braces, and would be quite inconsistent with the rest of the language.

**Comment by alrz** [comment]
>You're last example doesn't work unless lock (x) expression is also considered an expression,

that's what this issue is about.

> than just allowing the removal of the braces

it's already allowed. (it uses an embedded-statement)

>would be quite inconsistent with the rest of the language.

how?

**Comment by HaloFour** [comment]
@alrz 

> that's what this issue is about.

Oh yeah.  Sometimes one gets wrapped up in a new set of comments that you forget what the original proposal was about.

> > would be quite inconsistent with the rest of the language.

> how?

How many other control statements in the language followed by an embedded statement are expressions or can be expressions?

**Comment by alrz** [comment]
>How many other control statements in the language followed by an embedded statement

This would be a completely separate syntax form, just like https://github.com/dotnet/csharplang/issues/101 (btw, lock and using are not a control statement.)

**Comment by svick** [comment]
@HaloFour 

> How many other control statements in the language followed by an embedded statement are expressions or can be expressions?

There is a precedent for having an expression version of that kind of statement: ternary expression for `if` statement and now `switch` expression for `switch` statement (though that one is technically not an embedded statement).

In both those cases, the expression version has different syntax than the statement version, but I don't think there's a reason for that here.

The language is moving to be more expression-friendly and this would be another step in that direction.

**Comment by alrz** [comment]
with introduction of switch expressions, the need for something like "block expression" is more visible. while switch expressions and :? have specialized syntax and are extremely common, for others I think we don't need a new syntax form, we should just be able to wrap statements inside some kind of "block expression".

**Comment by lostmsu** [comment]
@HaloFour is there a scenario when it is hard to disambiguate `using(x) ...` statement from expression?
When parsing statements seeing a `using` token in the beginning means statement otherwise it would always be an expression.

**Comment by CyrusNajmabadi** [comment]
> otherwise it would always be an expression.

A core problem is that 'otherwise' case comes up a *lot* during editing :)

**Comment by lostmsu** [comment]
@CyrusNajmabadi are you saying it would slow down typing performance in IDE around `using` keyword?

**Comment by CyrusNajmabadi** [comment]
It's not a matter of perf. It's a matter of ambiguity and error recovery during editing scenarios.
